File: frontend/lib/model/expense.dart
/// Modelo de gasto. Ref: Solution Design, DB Schema p.12 — tabla EXPENSE.
/// Relaciones: user_id -> Users (N:1), category_id -> Categories (N:1).
class Expense {
  final int expenseId;
  final int userId;
  final int categoryId;
  final double amount;
  final String description;
  final DateTime expenseDate;
  final DateTime createdAt;
  final String? categoryName;

  const Expense({
    required this.expenseId,
    required this.userId,
    required this.categoryId,
    required this.amount,
    required this.description,
    required this.expenseDate,
    required this.createdAt,
    this.categoryName,
  });

  factory Expense.fromJson(Map<String, dynamic> json) => Expense(
        expenseId: json['expense_id'] as int,
        userId: json['user_id'] as int,
        categoryId: json['category_id'] as int,
        amount: double.parse(json['amount'].toString()),
        description: json['description'] as String,
        expenseDate: DateTime.parse(json['expense_date'] as String),
        createdAt: DateTime.parse(json['created_at'] as String),
        categoryName: json['category_name'] as String?,
      );

  Map<String, dynamic> toJson() => {
        'amount': amount,
        'description': description,
        'expense_date': expenseDate.toIso8601String().split('T').first,
        'category_id': categoryId,
      };
}

File: backend/app/models/expense.py
"""Modelo Expense: gasto individual registrado por un usuario.
Referencia: Solution Design, ERD p.11, DB Schema p.12 — tabla EXPENSE.
Relaciones: user_id -> Users (N:1), category_id -> Categories (N:1).
"""

from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import Date, DateTime, ForeignKey, Integer, Numeric, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Expense(Base):
    __tablename__ = "expenses"

    expense_id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.user_id"), nullable=False)
    category_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("categories.category_id"), nullable=False
    )
    amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    description: Mapped[str] = mapped_column(String(255), nullable=False)
    expense_date: Mapped[date] = mapped_column(Date, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime, nullable=False, default=func.now()
    )

    user = relationship("User", back_populates="expenses")
    category = relationship("Category", back_populates="expenses")
